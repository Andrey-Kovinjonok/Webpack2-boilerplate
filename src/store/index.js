export createClientStore from './StoreCreator.js';
export * as storageAct from './storageActions.js';
