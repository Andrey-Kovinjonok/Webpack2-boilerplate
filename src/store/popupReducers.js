
import { setField } from '../utils/otm';

const initialState = { };

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case 'CHANGE_DATA': {
      const d = setField(state, action.treePath, action.value);
      // console.log(d);
      return d;
    }

    case 'INIT_DATA': {
      const initData = action.data || initialState;
      return { isShowingPopup: initData.isShowingPopup || false };
    }

    case 'READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE': {
      const { isShowingPopup } = action.value || {};
      if (isShowingPopup === undefined) {
        return state;
      }

      return {
        ...state.popupReducer,
        isShowingPopup,
      };
    }

    default:
      return state || initialState;
  }
}
