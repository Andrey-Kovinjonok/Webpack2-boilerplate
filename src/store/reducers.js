import storageReducer from './storageReducers';
import popupReducer from './popupReducers';

export default { storageReducer, popupReducer };
