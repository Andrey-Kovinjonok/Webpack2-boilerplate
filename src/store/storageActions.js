import { getField } from '../utils/otm';

export function initData(data) {
  return {
    data,
    type: 'INIT_DATA',
  };
}

export function storeData(data) {
  return {
    ...data,
    type: 'CHANGE_DATA',
  };
}

export function getData(state, path) {
  return getField(state.storageReducer, [...path]);
}

export function addMessage(messageText) {
  return {
    value: messageText,
    type: 'ADD_MESSAGE',
  };
}

export function readMessagesAndOrChangePopupStatus(params) {
  return {
    value: params,
    type: 'READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE',
  };
}
