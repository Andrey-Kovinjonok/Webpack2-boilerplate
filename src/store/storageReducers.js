
import { setField } from '../utils/otm';

const initialState = { };

export default function storeReducer(state = initialState, action) {
  switch (action.type) {
    case 'CHANGE_DATA': {
      const d = setField(state, action.treePath, action.value);
      // console.log(d);
      return d;
    }

    case 'INIT_DATA': {
      const initData = action.data || initialState;
      return { data: { messages: initData.messages || [] } };
    }

    case 'ADD_MESSAGE': {
      const newId =
        (state.data.messages && state.data.messages.length > 0)
        ? state.data.messages[0].id + 1
        : 0;

      const messagesSnapshot = [
        {
          id: newId,
          title: action.value,
          datetime: new Date(),
          unread: true,
        },
        ...state.data.messages,
      ];
      const d = setField(state, ['data', 'messages'], messagesSnapshot);
      // console.log(d);
      return d;
    }

    case 'READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE': {
      const { numberReaded } = action.value || {};
      if (numberReaded === undefined) {
        console.log('state', state);
        return state;
      }
      console.log('action.value', action.value);

      const messages = (state.data && state.data.messages) ? state.data.messages : [];
      const messagesSnapshot = messages.reduce((accParams, message) => {
        const { acc, processedIndex } = accParams;
        if (processedIndex >= 0) {
          return {
            acc: [
              ...acc,
              {
                ...message,
                unread: false,
              },
            ],
            processedIndex: processedIndex - 1,
          };
        }

        return {
          acc: [
            ...acc,
            message,
          ],
          processedIndex,
        };
      }, {
        acc: [],
        // -1 mean we need mark if for all items need to set readed flag
        processedIndex: (numberReaded === -1) ? messages.length : numberReaded,
      });

      const d = setField(state, ['data', 'messages'], messagesSnapshot.acc);
      // console.log(d);
      return d;
    }

    default:
      return state || initialState;
  }
}
