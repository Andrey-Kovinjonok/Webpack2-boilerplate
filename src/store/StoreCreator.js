import { createStore, applyMiddleware, combineReducers } from 'redux';
import CatchPromise from 'redux-catch-promise';

import reducers from './reducers';

const thunkMiddleware = CatchPromise();

export default function createApiClientStore(initialState) {
  const store = createStore(
    combineReducers(reducers),
    initialState,
    applyMiddleware(thunkMiddleware),
  );

  if (module.hot) {
    module.hot.accept('./reducers', () => {
      const nextReducers = require('./reducers');
      store.replaceReducer(nextReducers);
    });
  }

  return store;
}
