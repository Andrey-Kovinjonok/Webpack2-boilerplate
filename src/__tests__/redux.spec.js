import { storageAct } from './../store';
import storageReducers from './../store/storageReducers';
import popupReducers from './../store/popupReducers';

test('INIT storage', () => {
  const testInitActionStorage =
    storageAct.initData({
      treePath: ['data', 'someFlag'],
      value: true,
    });
  expect(testInitActionStorage).toMatchSnapshot();

  const testInitReducer1 = storageReducers({}, {
    type: 'INIT_DATA',
    data: undefined,
  });
  expect(testInitReducer1).toMatchSnapshot();
  expect(testInitReducer1)
  .toEqual({
    data: {
      messages: [],
    },
  });

  const testInitReducer2 = storageReducers({}, {
    type: 'INIT_DATA',
    data: {},
  });
  expect(testInitReducer2).toMatchSnapshot();
  expect(testInitReducer2)
  .toEqual({
    data: {
      messages: [],
    },
  });

  const testInitReducer3 = storageReducers({}, {
    type: 'INIT_DATA',
    data: { messages: [{ id: 0, text: 'test' }] },
  });
  expect(testInitReducer3)
  .toEqual({
    data: {
      messages: [{ id: 0, text: 'test' }],
    },
  });
  expect(testInitReducer3).toMatchSnapshot();
});

test('CHANGE_DATA storage', () => {
  /*
  const initedStore = storageReducers({}, {
    type: 'INIT_DATA',
    data: undefined,
  });*/
  const changeDataResult = storageReducers({}, {
    type: 'CHANGE_DATA',
    treePath: ['data', 'messages'],
    value: [],
  });

  expect(changeDataResult.data).not.toBeNull();
  expect(changeDataResult.data).not.toBeUndefined();
  expect(changeDataResult.data.messages).not.toBeUndefined();
  expect(changeDataResult.data.messages.length).toBe(0);
});

test('ADD_MESSAGE', () => {
  const addMessageActionTest = storageAct.addMessage('SOME TEXT MESSAGE');
  expect(addMessageActionTest).toMatchSnapshot();

  const initedStore = storageReducers({}, {
    type: 'INIT_DATA',
    data: undefined,
  });
  const addFirstMessageResult = storageReducers(initedStore, {
    type: 'ADD_MESSAGE',
    value: 'SOME TEXT MESSAGE',
  });

  expect(addFirstMessageResult.data).not.toBeNull();
  expect(addFirstMessageResult.data).not.toBeUndefined();
  expect(addFirstMessageResult.data.messages).not.toBeUndefined();
  expect(addFirstMessageResult.data.messages.length).toBe(1);
  expect(addFirstMessageResult.data.messages[0]).not.toBeUndefined();
  expect(addFirstMessageResult.data.messages[0].id).toBe(0);
});

test('READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE storage', () => {
  const initedStore = storageReducers({}, {
    type: 'INIT_DATA',
    data: undefined,
  });
  const addFirstMessageResult = storageReducers(initedStore, {
    type: 'ADD_MESSAGE',
    value: 'SOME TEXT MESSAGE',
  });

  const undefinedParams = storageAct.readMessagesAndOrChangePopupStatus(undefined);
  expect(undefinedParams).toMatchSnapshot();

  const definedParams = storageAct.readMessagesAndOrChangePopupStatus({ numberReaded: 5 });
  expect(definedParams).toMatchSnapshot();


  const readOrPopupResult1 = storageReducers(addFirstMessageResult, {
    type: 'READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE',
    value: undefined,
  });
  expect(readOrPopupResult1.data.messages[0].unread).toEqual(true);

  const readOrPopupResult2 = storageReducers(addFirstMessageResult, {
    type: 'READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE',
    value: { numberReaded: 5 },
  });
  expect(readOrPopupResult2.data.messages[0].unread).toEqual(false);
});

test('INIT popup', () => {
  const testInitActionPopup =
    storageAct.initData({
      data: { isShowingPopup: true },
    });
  expect(testInitActionPopup).toMatchSnapshot();

  const testInitReducer1 = popupReducers({}, {
    type: 'INIT_DATA',
    data: undefined,
  });
  expect(testInitReducer1)
  .toEqual({
    isShowingPopup: false,
  });
  expect(testInitReducer1).toMatchSnapshot();

  const testInitReducer2 = popupReducers({}, {
    type: 'INIT_DATA',
    data: { isShowingPopup: true },
  });
  expect(testInitReducer2)
  .toEqual({
    isShowingPopup: true,
  });
  expect(testInitReducer2).toMatchSnapshot();
});

test('CHANGE_DATA popup', () => {
  const changeDataResult1 = storageReducers({}, {
    type: 'CHANGE_DATA',
    treePath: ['isShowingPopup'],
    value: true,
  });

  expect(changeDataResult1).not.toBeNull();
  expect(changeDataResult1).not.toBeUndefined();
  expect(changeDataResult1.isShowingPopup).not.toBeUndefined();
  expect(changeDataResult1.isShowingPopup).toBe(true);
});

test('READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE popup', () => {
  const initedStore = storageReducers({}, {
    type: 'INIT_DATA',
    data: undefined,
  });

  const undefinedParams = storageAct.readMessagesAndOrChangePopupStatus(undefined);
  expect(undefinedParams).toMatchSnapshot();

  const definedParams = storageAct.readMessagesAndOrChangePopupStatus({ isShowingPopup: true });
  expect(definedParams).toMatchSnapshot();


  const readOrPopupResult1 = popupReducers(initedStore, {
    type: 'READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE',
    value: undefined,
  });
  expect(readOrPopupResult1.isShowingPopup).toEqual(undefined);

  const readOrPopupResult2 = popupReducers(initedStore, {
    type: 'READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE',
    value: { isShowingPopup: false },
  });
  expect(readOrPopupResult2).toMatchSnapshot();
  expect(readOrPopupResult2.isShowingPopup).toEqual(false);
});
