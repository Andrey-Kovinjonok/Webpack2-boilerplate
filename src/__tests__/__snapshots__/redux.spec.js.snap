exports[`test ADD_MESSAGE 1`] = `
Object {
  "type": "ADD_MESSAGE",
  "value": "SOME TEXT MESSAGE",
}
`;

exports[`test INIT popup 1`] = `
Object {
  "data": Object {
    "data": Object {
      "isShowingPopup": true,
    },
  },
  "type": "INIT_DATA",
}
`;

exports[`test INIT popup 2`] = `
Object {
  "isShowingPopup": false,
}
`;

exports[`test INIT popup 3`] = `
Object {
  "isShowingPopup": true,
}
`;

exports[`test INIT storage 1`] = `
Object {
  "data": Object {
    "treePath": Array [
      "data",
      "someFlag",
    ],
    "value": true,
  },
  "type": "INIT_DATA",
}
`;

exports[`test INIT storage 2`] = `
Object {
  "data": Object {
    "messages": Array [],
  },
}
`;

exports[`test INIT storage 3`] = `
Object {
  "data": Object {
    "messages": Array [],
  },
}
`;

exports[`test INIT storage 4`] = `
Object {
  "data": Object {
    "messages": Array [
      Object {
        "id": 0,
        "text": "test",
      },
    ],
  },
}
`;

exports[`test READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE popup 1`] = `
Object {
  "type": "READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE",
  "value": undefined,
}
`;

exports[`test READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE popup 2`] = `
Object {
  "type": "READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE",
  "value": Object {
    "isShowingPopup": true,
  },
}
`;

exports[`test READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE popup 3`] = `
Object {
  "isShowingPopup": false,
}
`;

exports[`test READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE storage 1`] = `
Object {
  "type": "READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE",
  "value": undefined,
}
`;

exports[`test READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE storage 2`] = `
Object {
  "type": "READ_MESSAGES_AND_OR_CHANGE_POPUP_STATE",
  "value": Object {
    "numberReaded": 5,
  },
}
`;
